### PR TITLE
20241125 token user override

### PR DESCRIPTION
**Pull request by peterthomassen**
This implements #886.

### CODE REVIEW

File: api/desecapi/migrations/0041_rename_user_token_owner_token_user_override_and_more.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                related_name="+",
+                to=settings.AUTH_USER_MODEL,
+            ),
+        ),
+        migrations.AddField(
+            model_name="token",
+            name="user_override",
+            field=models.ForeignKey(
+                null=True,
+                on_delete=django.db.models.deletion.CASCADE,
```
Comment (nils-wisiol) [inline api/desecapi/migrations/0041_rename_user_token_owner_token_user_override_and_more.py:48]: add documentation

File: api/desecapi/migrations/0041_rename_user_token_owner_token_user_override_and_more.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            ),
+        ),
+        migrations.AddConstraint(
+            model_name="token",
+            constraint=models.UniqueConstraint(
+                fields=("id", "user_id"), name="unique_id_user"
+            ),
+        ),
+        migrations.RunSQL(
+            sql="ALTER TABLE desecapi_tokendomainpolicy ADD FOREIGN KEY ( token_id, token_user_id ) REFERENCES desecapi_token ( id, user_id );",
```
Comment (nils-wisiol) [inline api/desecapi/migrations/0041_rename_user_token_owner_token_user_override_and_more.py:83]: how does the name match up with the dropped constraint above?

File: api/desecapi/views/tokens.py [hunk 3]
```diff
@@ -88,7 +88,7 @@ def create(self, request, *args, **kwargs):
             raise Http404
 
     def get_queryset(self):
-        qs = Token.objects.filter(user=self.request.user)
+        qs = Token.objects.filter(owner=self.request.user)
```
Comment (nils-wisiol) [inline api/desecapi/views/tokens.py:91]: `user=request.user or owner=request.user`?

Comment (peterthomassen) [reply api/desecapi/views/tokens.py:91]: At the time of this commit, which does not expose new functionality, it should be `owner`. Subsequent commits adjust this as functionality is changed.

File: api/desecapi/migrations/0041_rename_user_token_owner_token_user_override_and_more.py (hunk 2 above)
Comment (peterthomassen) [reply api/desecapi/migrations/0041_rename_user_token_owner_token_user_override_and_more.py:83]: The name is generated automatically, and I've verified that it does match up.

File: api/desecapi/migrations/0041_rename_user_token_owner_token_user_override_and_more.py (hunk 1 above)
Comment (peterthomassen) [reply api/desecapi/migrations/0041_rename_user_token_owner_token_user_override_and_more.py:48]: Will do. For the record:

The "naive" implementation gives the behavior that
- deletion of the impersonated user also deletes the impersonation token (insead of returning it do its owner). This is due to `on_delete=CASCADE`;
- to detach an impersonation token from the target account, the target (effective) user can simply delete it. This is "the normal behavior" for token management.

Impersonation tokens can only be managed by the target (effective) user, not by their owner. Both users however can see the token and its configuration.

We will expose `user_override` in the token serializer, and will allow writing an account identifier (likely the email address) into it in order to request approval from the target user. We need to ensure that the only allowed transition is "from NULL to user" (but not "from user to NULL" or "from user1 to user2").

Shall document all of this.

**Comment by peterthomassen** [comment]
View adjustments, tests, and docs done -- ready for review.

(Might add some more tests, e.g.: setting `user_override` on `POST tokens/` has no effect; make a DNS change with an override token; access an override token's policy from owner or override account.)

File: api/desecapi/migrations/0041_rename_user_token_owner_token_user_override_and_more.py (hunk 1 above)
Comment (peterthomassen) [reply api/desecapi/migrations/0041_rename_user_token_owner_token_user_override_and_more.py]: done.

**Comment by peterthomassen** [comment]
Final part has been added, with the workflow from #858 now fully functioning.

Missing:
- [ ] nicer text (warnings) in email template and activation GUI
- [ ] tests for 237ebd9b98b7dca166d71a71cf572fd9e3601c41 (will be added as a fixup)
- [ ] docs (will not document for now; prefer to add in January or so when we know the API is stable)

File: api/desecapi/permissions.py [hunk 4]
```diff
@@ -67,6 +67,25 @@ def has_object_permission(self, request, view, obj):
...
+    Custom permission to only allow object access for the user on record.
+    """
+
+    def has_object_permission(self, request, view, obj):
+        return obj.user == request.user
+
+
+class IsTokenUser(permissions.BasePermission):
+    """
+    Custom permission to only allow object access for the user on record.
```
Comment (nils-wisiol) [inline api/desecapi/permissions.py:81]: stale doc

File: docs/auth/tokens.rst [hunk 5]
```diff
@@ -482,6 +501,55 @@ and you cannot remove a default policy when other policies are still in place.
...
+
+When listing tokens, the configuration of override tokens is visible to both
+their owner and the user listed in the ``user_override`` field.
+To discern which tokens are (or are not) override tokens, associated email
+addresses are listed in both the ``owner`` and ``user_override`` fields.
+Note that as a result, both parties can observe when either deSEC account email
+address changes.
+
+Only API tokens without the ``perm_manage_tokens`` permission are eligible to
+become override tokens.
```
Comment (nils-wisiol) [inline docs/auth/tokens.rst:538]: Why?

File: docs/auth/tokens.rst [hunk 6]
```diff
@@ -482,6 +501,55 @@ and you cannot remove a default policy when other policies are still in place.
...
+Further, only the override user (not the token owner) can manage an override
+token, including changing its name or permissions.
+However, both the owner and the override user can delete an override token.
```
Comment (nils-wisiol) [inline docs/auth/tokens.rst:541]: We may need a UI warning for this

File: api/desecapi/views/domains.py [hunk 7]
```diff
@@ -80,6 +81,18 @@ def domain(self):
     def get_queryset(self):
         qs = self.request.user.domains
 
+        if (  # other conditions? check policies?
```
Comment (nils-wisiol) [inline api/desecapi/views/domains.py]: revisit

File: api/desecapi/views/tokens.py [hunk 8]
```diff
@@ -16,15 +17,24 @@
...
+            permissions.HasManageTokensPermission,
+        ]
+        # The effective user may manage the token; its owner can only delete it
+        if self.request.method not in SAFE_METHODS and self.action != "destroy":
+            ret.append(permissions.IsUser)
+        return ret
+
     def get_queryset(self):
-        return self.request.user.token_set.all()
+        return Token.objects.filter(
```
Comment (nils-wisiol) [inline api/desecapi/views/tokens.py:35]: revisit

**Review by nils-wisiol — COMMENTED** [review]
Some comments but haven't looked at all commits yet. Commented 'revisit' for things we need to look at together.

File: api/desecapi/views/domains.py (hunk 7 above)
Comment (peterthomassen) [reply api/desecapi/views/domains.py]: Filter regardless of `auto_policy`, only allow reading domains for which policies allow any kind of permission; if no policy exists, all domains are visible

File: api/desecapi/permissions.py (hunk 4 above)
Comment (peterthomassen) [reply api/desecapi/permissions.py:81]: done

File: docs/auth/tokens.rst (hunk 5 above)
Comment (peterthomassen) [reply docs/auth/tokens.rst:543]: done

File: docs/auth/tokens.rst (hunk 6 above)
Comment (peterthomassen) [reply docs/auth/tokens.rst:546]: done

image

File: api/desecapi/views/authenticated_actions.py [hunk 9]
```diff
@@ -196,8 +196,7 @@ def post(self, request, *args, **kwargs):
         return Response(
             {
                 "detail": f"Success! Welcome to deSEC. {self.authenticated_action.token.owner.email} will "
-                "manage your domains for you. Should you want to access your account directly, you can do "
-                "so at any time. (Remember to first set a password for your account.)"
+                "manage your domains for you. Set a password any time to access your account directly."
```
Comment (nils-wisiol) [inline api/desecapi/views/authenticated_actions.py:199]: "Reset the password to your account at any time to access your account directly."

**Review by nils-wisiol — APPROVED** [review]
lgtm, three nits found

File: docs/auth/tokens.rst [hunk 10]
```diff
@@ -539,10 +539,11 @@ addresses are listed in both the ``owner`` and ``user_override`` fields.
 Note that as a result, both parties can observe when either deSEC account email
 address changes.
 
-Only API tokens without the ``perm_manage_tokens`` permission are eligible to
-become override tokens.
-Further, only the override user (not the token owner) can manage an override
-token, including changing its name or permissions.
+Only the override user (not the token owner) can manage an override token,
+including changing its name or permissions.
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply docs/auth/tokens.rst:543]: Yes. I've told our sponsor explicitly. We might want to handle this better in the future (but I currently have no good idea how).

[the remark this answers is not in the document]
Comment (peterthomassen) [reply docs/auth/tokens.rst:543]: done

File: api/desecapi/templates/emails/activate-account-with-override-token/content.txt [hunk 11]
```diff
@@ -1,26 +1,19 @@
 {% extends "emails/content.txt" %}
 {% block content %}{% load action_extras %}Hi there,
 
-TODO
+{{ action_serializer.instance.token.owner.email }} has requested that you open a deSEC account in order to manage DNS records for your domains.
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/templates/emails/activate-account-with-override-token/content.txt:4]: fixed similarly

File: api/desecapi/views/authenticated_actions.py (hunk 9 above)
Comment (peterthomassen) [reply api/desecapi/views/authenticated_actions.py:199]: done

**Pull request merged**